packages/frontend/src/components/rich_text_editor/rich_text_editor.tsx

```diff
@@ -0,0 +1,360 @@
+        isBoldActive: false,
+        isEmActive: false,
+    });
+
+    const [headingLevel, setHeadingLevel] = createSignal<number | null>(null);
+    const isReady = useDocHandleReady(() => props.handle);
+
+    createEffect(() => {
+        // NOTE: Make the effect depend on the given ID to ensure that this
+        // component updates when the Automerge handle and path both stay the
+        // same but the path refers to a different object in the document.
+        props.id;
+
+        if (!isReady()) {
+            return;
+        }
+
+        const { schema, pmDoc, automergePlugin } = proseMirrorAutomergeInit(
+            props.handle,
+            props.path,
+        );
+
+        const blockMathInputRule = makeBlockMathInputRule(
+            REGEX_BLOCK_MATH_DOLLARS,
+            schema.nodes.math_display,
+        );
+
+        const plugins: Plugin[] = [
+            keymap(richTextEditorKeymap(schema, props)),
+            keymap(baseKeymap),
+            ...(props.placeholder ? [initPlaceholderPlugin(props.placeholder)] : []),
+            automergePlugin,
+            mathPlugin,
+            inputRules({ rules: [blockMathInputRule] }),
+        ];
+
+        const state = EditorState.create({ schema, plugins, doc: pmDoc });
+        const view = new EditorView(editorRoot, {
+            state,
+            dispatchTransaction: (tx: Transaction) => {
+                // XXX: It appears that automerge-prosemirror can dispatch
+                // transactions even after the view has been destroyed.
+                if (view.isDestroyed) {
+                    return;
+                }
+
+                const newState = view.state.apply(tx);
+                setHeadingLevel(activeHeading(view.state, schema));
+                setMarkStates(activeMarks(newState, schema));
+                view.updateState(newState);
+            },
+            handleDOMEvents: {
+                focus: () => {
+                    setEditorFocused(true);
+                    props.onFocus?.();
+                    return false;
+                },
+                blur: (_, event) => {
+                    const relatedTarget = (event as FocusEvent).relatedTarget as Node | null;
+
+                    if (
+                        relatedTarget &&
+                        (menuRoot.contains(relatedTarget) || editorRoot.contains(relatedTarget))
+                    ) {
+                        return true;
+                    }
+
+                    setEditorFocused(false);
+                    return false;
+                },
+            },
+            clipboardTextSerializer: (slice) => {
+                return mathSerializer.serializeSlice(slice);
+            },
+        });
+
+        if (props.ref) {
+            props.ref(view);
+        }
+
+        setHeadingLevel(activeHeading(view.state, schema));
+        setMarkStates(activeMarks(view.state, schema));
+
+        setMenuControls({
```

Comment:
This method for passing callbacks to the menu is relatively weird, however I've seen a couple examples of it and this may be the least messy way of dealing with the asynchronous loading of the document.